miden-tx/src/host/tx_authenticator.rs

```diff
@@ -0,0 +1,153 @@
+use alloc::vec::Vec;
+
+use miden_objects::{
+    accounts::AccountDelta,
+    crypto::dsa::rpo_falcon512::{Polynomial, SecretKey},
+};
+use rand::Rng;
+use vm_processor::{Felt, Word};
+
+use crate::error::AuthenticationError;
+
+// TRANSACTION AUTHENTICATOR
+// ================================================================================================
+
+/// Represents an authenticator for transactions.
+///
+/// Its main use is to provide a method to create a DSA signature for a message,
+/// based on an [AccountDelta].
+///
+/// The signature is intended to provide authentication. The implementer can verify
+/// and approve any changes before the message is signed.
+pub trait TransactionAuthenticator {
+    /// Retrieves a signataure for a specific message as a list of [Felt].
+    /// The request is initiaed by the VM as a consequence of the SigToStack advice
+    /// injector.
+    ///
+    /// - `pub_key`: The public key used for signature generation.
+    /// - `message`: The message to sign, usually a commitment to the transaction data.
+    /// - `account_delta`: An informational parameter describing the changes made to
+    ///   the account up to the point of calling `get_signature()`. This allows the
+    ///   authenticator to review any alterations to the account prior to signing.
+    ///   It should not be directly used in the signature computation.
+    fn get_signature(
+        &mut self,
+        pub_key: Word,
+        message: Word,
+        account_delta: &AccountDelta,
+    ) -> Result<Vec<Felt>, AuthenticationError>;
+}
+
+// FALCON AUTHENTICATOR
+// ================================================================================================
+
+/// Represents a signer for Falcon signatures, based on a user's [SecretKey]
+#[derive(Clone, Debug)]
+pub struct FalconAuthenticator<R> {
+    secret_key: SecretKey,
+    rng: R,
+}
```

Comment:
One thing we can do is make this a bit more generic and call it something like `BasicAuthenticator`. It could look something like this:
```Rust
pub enum SecretKey {
    RpoFalcon512(rpo_falcon512::SecretKey),
}

pub struct BasicAuthenticator<R> {
    /// pub_key |-> secret_key mapping
    key: BTreeMap<Word, SecretKey>,
    rng: R,
}

impl<R: Rng> BasicAuthenticator<R> {
    pub fn new(keys: &[(Word, SecretKey)]) -> Self {
        todo!()
    }
}
```
The benefits of this approach are:
- We can support multiple keys in the authenticator (even multiple keys of the same kind).
- It is pretty easy to extend this scheme with additional signatures.

Reply:
The only problem with this is that `Word` is not `Ord` so it cannot be a key for a `BTreeMap`. Is using `Digest` safe for this?

Reply:
Yep - we can use `Digest` here.